src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::BoolT,
+                            actual: inner_type,
+                        })
+                    }
+                }
+                UOp::Negate => {
+                    if let Type::NumT(range) = inner_type {
+                        Ok(Type::NumT(range))
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::NumT(0..1),
+                            actual: inner_type,
+                        })
+                    }
+                }
+            }
+        }
+        // Casting is always of the type of the inner expression
+        Expr::Cast {
+            inner,
+            typ,
+            overflow: _overflow,
+        } => {
+            // First, type check the inner expression
+            let inner_type = typeck_expr(inner, ctx)?;
+
+            // According to the [T-Cast] rule, both the inner expression and the target type
+            // must be numeric types with bounds
+            match (&inner_type, typ) {
+                (Type::NumT(_), Type::NumT(_)) => {
+                    // The cast is valid, return the target type
+                    Ok(typ.clone())
+                }
+                (_, Type::NumT(_)) => {
+                    // The inner expression is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+                (Type::NumT(_), _) => {
+                    // The target type is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: typ.clone(),
+                    })
+                }
+                _ => {
+                    // Neither the inner expression nor the target type are numeric types
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+            }
+        }
+        // Call expressions are always of the type of the function
+        Expr::Call { callee, args } => {
+            // Type check each argument
+            let mut arg_types: Vec<Type> = Vec::new();
+            for arg in args {
+                arg_types.push(typeck_expr(arg, ctx)?);
+            }
+
+            // Look up the function in the function environment
+            let function = ctx
+                .funcs
+                .get(callee)
+                .ok_or(TypeError::UndefinedFunction(*callee))?;
+
+            // Check that the number of arguments matches the number of parameters
+            if arg_types.len() != function.params.len() {
```

Comment:
I suggest not having different type errors and having a single check so you can also get rid of the `for` loop after this `if` statement (something like `arg_types == function.params.iter().map(|(_, t)| t.clone()).collect()`)